### PR TITLE
(Bug) #2409 (Bug) dashboard pie charts show incorrect number of users

### PR DESCRIPTION
**Pull request by yaroslav-fedyshyn-nordwhale**
# Description

The account with zero balance wasn't included in the distributed histogram.
- include zero balance accounts to the result from the server.

About https://app.zenhub.com/workspaces/gooddollar-5d50833888a83c6880d3e345/issues/gooddollar/gooddapp/2409undefined

# How Has This Been Tested?

The summary of the pie chart sections should equal the number of existed accounts.

# Checklist:
- [x] PR title matches follow: (Feature|Bug|Chore) Task Name
- [x] My code follows the style guidelines of this project
- [x] I have followed all the instructions described in the initial task (check Definitions of Done)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added reference to a related issue in the repository
- [x] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [x] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: server/src/providers/wallets.ts [hunk 1]
```diff
@@ -198,16 +198,16 @@ class Wallets {
...
-        key = `${ceilFunc(min+1)}-${maxStep}`
+        filter[field] = { $gte: 0, $lte: maxStep }
+        key = `${ceilFunc(min)}-${maxStep}`
       } else if (j>=step-1) {
-        filter[field] = { $gt: minStep, $lte: max }
+        filter[field] = { $gte: minStep, $lte: max }
         key = `${ceilFunc(minStep+1)}-${max}`
       }
-
+      console.log('qqq', filter)
```
Comment (serdiukov-o-nordwhale) [inline server/src/providers/wallets.ts]:
```suggestion
```

File: server/src/providers/wallets.ts [hunk 2]
```diff
@@ -198,16 +198,16 @@ class Wallets {
...
 
       const filter = {
-        [field]:{ $gt: minStep, $lte: maxStep }
+        [field]:{ $gte: minStep, $lte: maxStep }
       }
       if (j===0) {
-        filter[field] = { $gt: 0, $lte: maxStep }
-        key = `${ceilFunc(min+1)}-${maxStep}`
+        filter[field] = { $gte: 0, $lte: maxStep }
+        key = `${ceilFunc(min)}-${maxStep}`
```
Comment (serdiukov-o-nordwhale) [inline server/src/providers/wallets.ts:206]: why '+1' was removed here but kept in 'minStep+1' ?

**Review by serdiukov-o-nordwhale — CHANGES_REQUESTED** [review]

**Comment by sirpy** [comment]
@yaroslav-fedyshyn-nordwhale it doesnt make sense, we have at the moment 1600 unique claimers that means at least 1600 users have balance >0 and the pie charts do not reflect that

File: server/src/providers/wallets.ts (hunk 2 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply server/src/providers/wallets.ts:206]: it shouldn't be removed as the $gte used only for the first step. See new commits

**Comment by yaroslav-fedyshyn-nordwhale** [comment]
> @yaroslav-fedyshyn-nordwhale it doesnt make sense, we have at the moment 1600 unique claimers that means at least 1600 users have balance >0 and the pie charts do not reflect that

ok. it is really strange.
I checked the prod database - and I see only 134 records in 'wallets' collection.
The total summary of the pie chart sections is 129. And also there are 5 records with a balance 0 (which is not included yet)
129 + 5 = 134 - so the query works fine (except zero balances, this PR should solve it).
I think the issue appears when storing/updating wallet balances

Maybe issue is here:
```
await this.tokenContract.getPastEvents('Transfer')
```

Does the 'Transfer' event contain 'Claim' events? 
Or it will return only events when users are sending/receiving G$s?
I think it's not, so it will be updating the wallet collection only for those users who did some G$ transfers, and not for those who claim.
So i think we need to update wallets also in the updateClaimEvents method.

**Comment by yaroslav-fedyshyn-nordwhale** [comment]
See the last commit - I think it should resolve the issue.

File: server/src/providers/blockchain.ts [hunk 3]
```diff
@@ -301,12 +307,8 @@ export class blockchain {
    * @return {Promise<void>}
    */
   async checkAddressesClaimed (arrayOfAddresses: string[]): Promise<void> {
-    // there could be duplicates, so need to get unique values
-    // new Set([...]) -> will return unique values from received array
-    const uniqueAddresses = [...new Set(arrayOfAddresses)]
-
     // check multiple addresses exists and create new records in case if not exist by one db query
-    const { nonExistedCount } = await AddressesClaimedProvider.checkIfExistsMultiple(uniqueAddresses)
+    const { nonExistedCount } = await AddressesClaimedProvider.checkIfExistsMultiple(arrayOfAddresses)
```
Comment (serdiukov-o-nordwhale) [inline server/src/providers/blockchain.ts:312]: 1. why you don't use uniqueAddresses  here ?
2. as we're using lodash, we could use it's uniq() function

Comment (serdiukov-o-nordwhale) [reply server/src/providers/blockchain.ts:312]: 3. wallet addresses are case-insensitive, so 0xABCDxxxx and 0xaBCcdxxxx are the same addresses. So i think the most better is to use uniqBy(arrayOfAddresses , toLower) (form lodash)

**Review by serdiukov-o-nordwhale — CHANGES_REQUESTED** [review]

File: server/src/providers/blockchain.ts (hunk 3 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply server/src/providers/blockchain.ts:312]: 1. I move uniq to the updateClaimEvents (line 380)
2. done
3. done

Comment (serdiukov-o-nordwhale) [reply server/src/providers/blockchain.ts:312]: 1. i saw that but why non-unique addresses are now passed to AddressesClaimedProvider.checkIfExistsMultiple() ? why you decided do not pass the unique ones like it was before ?

File: server/src/providers/blockchain.ts [hunk 4]
```diff
@@ -376,8 +377,7 @@ export class blockchain {
 
     if (allAddresses.length) {
       // there could be duplicates, so need to get unique values
-      // new Set([...]) -> will return unique values from received array
-      const uniqueAddresses = [...new Set(allAddresses)]
+      const uniqueAddresses = uniq(allAddresses)
```
Comment (serdiukov-o-nordwhale) [inline server/src/providers/blockchain.ts]: isn't uniqBy(allAddresses, toLower) working ?

Comment (yaroslav-fedyshyn-nordwhale) [reply server/src/providers/blockchain.ts]:
```
uniqBy(allAddresses, toLower)
```
It just checks the values being lower case applied but return the original value.

Comment (serdiukov-o-nordwhale) [reply server/src/providers/blockchain.ts]: @yaroslav-fedyshyn-nordwhale yep i would have exactly that behavior - to keep original values

Comment (yaroslav-fedyshyn-nordwhale) [reply server/src/providers/blockchain.ts]: done

**Review by serdiukov-o-nordwhale — APPROVED** [review]

**Pull request merged**